maps/Hydration/HDRUK/2.2.1/translation.jsonata

```diff
@@ -0,0 +1,221 @@
+            "enum_titles": $types.options
+        } :
+    $types.format ?
+        {
+            "type": $types.type,
+            "format": $types.format
+        } : null
+};
+
+$listValidation := function($component){
+    {
+        "required": $component.required,
+        "type": "array",
+        "of": $getItemsType($component.types)
+    }
+};
+
+$stringValidation := function($component){
+    {
+        "required": $component.required,
+        "type": $component.types.type,
+        "min": $component.types.minLength,
+        "max": $component.types.maxLength
+    }
+};
+
+$formatValidation := function($component){
+    {
+        "required": $component.required,
+        "type": $component.types.type,
+        "format": $component.types.format
+    }
+};
+
+$enumValidation := function($component){
+    {
+        "required": $component.required,
+        "type": $component.types.type,
+        "enum": $component.types.options,
+        "enum_titles": $component.types.options
+    }
+};
+
+$regexValidation := function($component){
+    {
+        "required": $component.required,
+        "type": $component.types.type,
+        "pattern": $component.types.pattern
+    }
+};
+
+$intValidation := function($component){
+    {
+        "required": $component.required,
+        "type": "integer",
+        "positive": true
+    }
+};
+
+$dateValidation := function($component){
+    {
+        "required": $component.required,
+        "type": $component.types
+    }
+};
+
+$validationSchema := function($component, $f){
+    {
+        "schema": {
+            "title": $component.title,
+            "type": "object",
+            "properties": {
+                "field": $f($component)
+            }
+        }
+    }
+};
+
+$get_validation := function($component){
+    $component.is_list ?
+        $validationSchema($component, $listValidation) :
+    $component.types.maxLength ?
+        $validationSchema($component, $stringValidation) :
+    $component.types.format ?
+        $validationSchema($component, $formatValidation) :
+    $component.types.options ?
+        $validationSchema($component, $enumValidation) :
+    $component.types.pattern ?
+        $validationSchema($component, $regexValidation) :
+    $component.types = "int" ? 
+        $validationSchema($component, $intValidation) :
+    $component.types = "date" ?
+        $validationSchema($component, $dateValidation) : null
+};
+
+$define_components := function($components){
+    $map($components, function($component){
+        {
+            "title": $component.title,
+            "description": $component.description,
+            "location": $component.location,
+            "guidance": $component.guidance ? $component.guidance : "",
+            "field": $get_component($component),
+            "validation": $get_validation($component)
+        }
+    })
+};
+
+{
+    "schema_fields": $define_components(schema_fields),
+    "array_forms": [
```

Comment:
This indicates the parts of the form which can be repeated, observations is the only one I saw on the design for now.

Reply:
just put a comment in for a suggestion that could be marginally better, as the FE would be looping over each section (each component). It would have to look up a "level up" in the structure to retrieve `array_forms` and then filter on that, whereas I think it'd be easier to know, while looping over the component, if it was an array form or not. 

I think it's fine either way! As long as the FE can pull down information on which of the sections is an array

Reply:
I think I see what you're getting at.  Just to clarify there are two "array" like cases to handle.  There's one where a single field can have multiple values entered (for that we use the `autoComplete` component). And there's the case where a whole set of fields can be repeated, as if the case with observations.

To handle this second case, I've made an update.  Each component will now have an additional key `is_array_form` which will be true if the component is part of a set of fields that case be repeated e.g. components on the `observations` page on the design.  Do you think this makes it easier on the FE?

Reply:
Yeah great.  I think that'd make it easier. Would be better for question bank to have the structure like this too, there it's way more likely to have forms and forms-within-forms (sub-forms) that need to be `is_array_form`